File: swim-java/swim-runtime/swim-core/swim.warp/src/main/java/swim/warp/Envelope.java
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.warp;

import swim.codec.Debug;
import swim.codec.Decoder;
import swim.codec.Encoder;
import swim.codec.Format;
import swim.codec.Output;
import swim.codec.Writer;
import swim.recon.Recon;
import swim.structure.Form;
import swim.structure.Value;
import swim.uri.Uri;

public abstract class Envelope implements Debug {

  Envelope() {
    // sealed
  }

  public abstract String tag();

  public abstract Form<? extends Envelope> form();

  public abstract Uri nodeUri();

  public abstract Uri laneUri();

  public abstract Value body();

  public abstract Envelope nodeUri(Uri node);

  public abstract Envelope laneUri(Uri lane);

  public abstract Envelope body(Value body);

  @SuppressWarnings("unchecked")
  public Value toValue() {
    return ((Form<Envelope>) this.form()).mold(this).toValue();
  }

  public Encoder<?, Envelope> reconEncoder() {
    return new EnvelopeEncoder(this);
  }

  public Writer<?, ?> reconWriter() {
    return Recon.write(Output.full(), this.toValue());
  }

  public Writer<?, ?> writeRecon(Output<?> output) {
    return Recon.write(output, this.toValue());
  }

  public String toRecon() {
    return Recon.toString(this.toValue());
  }

  @Override
  public abstract <T> Output<T> debug(Output<T> output);

  @Override
  public String toString() {
    return Format.debug(this);
  }

  private static Decoder<Envelope> decoder;

  public static Decoder<Envelope> decoder() {
    if (Envelope.decoder == null) {
      Envelope.decoder = new EnvelopeDecoder();
    }
    return Envelope.decoder;
  }

  private static Encoder<Envelope, Envelope> encoder;

  public static Encoder<Envelope, Envelope> encoder() {
    if (Envelope.encoder == null) {
      Envelope.encoder = new EnvelopeEncoder();
    }
    return Envelope.encoder;
  }

  public static Envelope fromValue(Value value) {
    final String tag = value.tag();
    final Form<? extends Envelope> form = form(tag);
    if (form != null) {
      return form.cast(value);
    } else {
      return null;
    }
  }

  public static Envelope parseRecon(String recon) {
    final Value value = Recon.parse(recon);
    return Envelope.fromValue(value);
  }

  @SuppressWarnings("unchecked")
  public static <E extends Envelope> Form<E> form(String tag) {
    if ("event".equals(tag)) {
      return (Form<E>) EventMessage.FORM;
    } else if ("command".equals(tag)) {
      return (Form<E>) CommandMessage.FORM;
    } else if ("link".equals(tag)) {
      return (Form<E>) LinkRequest.FORM;
    } else if ("linked".equals(tag)) {
      return (Form<E>) LinkedResponse.FORM;
    } else if ("sync".equals(tag)) {
      return (Form<E>) SyncRequest.FORM;
    } else if ("synced".equals(tag)) {
      return (Form<E>) SyncedResponse.FORM;
    } else if ("unlink".equals(tag)) {
      return (Form<E>) UnlinkRequest.FORM;
    } else if ("unlinked".equals(tag)) {
      return (Form<E>) UnlinkedResponse.FORM;
    } else if ("auth".equals(tag)) {
      return (Form<E>) AuthRequest.FORM;
    } else if ("authed".equals(tag)) {
      return (Form<E>) AuthedResponse.FORM;
    } else if ("deauth".equals(tag)) {
      return (Form<E>) DeauthRequest.FORM;
    } else if ("deauthed".equals(tag)) {
      return (Form<E>) DeauthedResponse.FORM;
    }
    return null;
  }

}

File: swim-js/swim-runtime/swim-host/swim-warp/src/main/Envelope.ts
// Copyright 2015-2023 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type {Value} from "@swim/structure";
import {Recon} from "@swim/recon";
import type {AnyUri, Uri} from "@swim/uri";
import {Message} from "./Message";
import {EventMessage} from "./"; // forward import
import {CommandMessage} from "./"; // forward import
import {LinkRequest} from "./"; // forward import
import {LinkedResponse} from "./"; // forward import
import {SyncRequest} from "./"; // forward import
import {SyncedResponse} from "./"; // forward import
import {UnlinkRequest} from "./"; // forward import
import {UnlinkedResponse} from "./"; // forward import
import {AuthRequest} from "./"; // forward import
import {AuthedResponse} from "./"; // forward import
import {DeauthRequest} from "./"; // forward import
import {DeauthedResponse} from "./"; // forward import

/** @public */
export abstract class Envelope<E extends Envelope<E> = Envelope<any>> extends Message<E> {
  abstract readonly node: Uri;

  abstract withNode(node: AnyUri): E;

  abstract readonly lane: Uri;

  abstract withLane(lane: AnyUri): E;

  declare readonly prio: number; // getter defined below to work around useDefineForClassFields lunacy

  withPrio(prio: number): E;
  withPrio(this: E, prio: number): E {
    return this;
  }

  declare readonly rate: number; // getter defined below to work around useDefineForClassFields lunacy

  withRate(rate: number): E;
  withRate(this: E, rate: number): E {
    return this;
  }

  static override fromValue(value: Value): Envelope | null {
    switch (value.tag) {
      case "event": return EventMessage.fromValue(value);
      case "command": return CommandMessage.fromValue(value);
      case "link": return LinkRequest.fromValue(value);
      case "linked": return LinkedResponse.fromValue(value);
      case "sync": return SyncRequest.fromValue(value);
      case "synced": return SyncedResponse.fromValue(value);
      case "unlink": return UnlinkRequest.fromValue(value);
      case "unlinked": return UnlinkedResponse.fromValue(value);
      case "auth": return AuthRequest.fromValue(value);
      case "authed": return AuthedResponse.fromValue(value);
      case "deauth": return DeauthRequest.fromValue(value);
      case "deauthed": return DeauthedResponse.fromValue(value);
      default: return null;
    }
  }

  static override parseRecon(input: string): Envelope | null {
    return this.fromValue(Recon.parse(input));
  }
}
Object.defineProperty(Envelope.prototype, "prio", {
  get(this: Envelope): number {
    return 0;
  },
  configurable: true,
});
Object.defineProperty(Envelope.prototype, "rate", {
  get(this: Envelope): number {
    return 0;
  },
  configurable: true,
});
